examples/chatbot/src/websocket.rs

```diff
@@ -0,0 +1,172 @@
+                let msg = match msg {
+                    Err(tungstenite::Error::Protocol(
+                        tungstenite::error::ProtocolError::ResetWithoutClosingHandshake,
+                    )) => {
+                        tracing::warn!(
+                            "connection was sent an unexpected frame or was reset, reestablishing it"
+                        );
+                        s = self
+                            .connect().instrument(span)
+                            .await
+                            .context("when reestablishing connection")?;
+                        continue
+                    }
+                    _ => msg.context("when getting message")?,
+                };
+                self.process_message(msg, &mut event_fn).instrument(span).await?
+            })
+        }
+    }
+
+    /// Process a message from the websocket
+    async fn process_message<Fut>(
+        &mut self,
+        msg: tungstenite::Message,
+        event_fn: &mut impl FnMut(Event, types::Timestamp) -> Fut,
+    ) -> Result<(), eyre::Report>
+    where
+        Fut: std::future::Future<Output = Result<(), eyre::Report>>,
+    {
+        match msg {
+            tungstenite::Message::Text(s) => {
+                tracing::trace!("{s}");
+                // Parse the message into a [twitch_api::eventsub::EventsubWebsocketData]
+                match Event::parse_websocket(&s)? {
+                    EventsubWebsocketData::Welcome {
+                        payload: WelcomePayload { session },
+                        ..
+                    }
+                    | EventsubWebsocketData::Reconnect {
+                        payload: ReconnectPayload { session },
+                        ..
+                    } => {
+                        self.process_welcome_message(session).await?;
+                        Ok(())
+                    }
+                    EventsubWebsocketData::Notification { metadata, payload } => {
+                        event_fn(payload, metadata.message_timestamp.into_owned()).await?;
+                        Ok(())
+                    }
+                    EventsubWebsocketData::Revocation {
+                        metadata,
+                        payload: _,
+                    } => eyre::bail!("got revocation event: {metadata:?}"),
+                    EventsubWebsocketData::Keepalive {
+                        metadata: _,
+                        payload: _,
+                    } => Ok(()),
+                    _ => Ok(()),
+                }
+            }
+            tungstenite::Message::Close(_) => todo!(),
+            _ => Ok(()),
+        }
+    }
+
+    async fn process_welcome_message(&mut self, data: SessionData<'_>) -> Result<(), eyre::Report> {
+        tracing::info!("connected to twitch chat");
+        self.session_id = Some(data.id.to_string());
+        if let Some(url) = data.reconnect_url {
+            self.connect_url = url.parse()?;
+        }
```

Comment:
If we have this reconnect URL, we subscribe to the topics again (and one more time when we receive a welcome message on the new socket). Not super sure, but does Twitch deduplicate the subs?

Reply:
I assumed they got dropped, but your assumption is correct. 

https://dev.twitch.tv/docs/eventsub/handling-websocket-events/#reconnect-message

> The connection will include the same subscriptions that the old connection had.